Make edition component responsive

koderjoker opened the pull request:
> **Description**: What does this PR achieve? [hotfix]

Makes the edition component responsive

Closes #1339

> **Evidence**: If this PR touches UI, please post evidence (screenshot) of it behaving correctly:

For laptops screens
<img src="https://user-images.githubusercontent.com/30471843/52917483-22ee9600-3312-11e9-9e00-bc899b0b95ba.png"  width="400"/>

For mobile
<img src="https://user-images.githubusercontent.com/30471843/52917484-241fc300-3312-11e9-95fe-4423101e2ba1.png"  width="200"/>

koderjoker commented:
Changes required:

- Change the use of editions id to follow linting rules
- Make sure the interface doesn't break at tablet breakpoint (768px) while maintaining mobile first approach
The interface breaks exactly at 768px. Most likely because of overlapping styles of editions.less (uses min-width) and common.less (uses max-width).
<img src="https://user-images.githubusercontent.com/30471843/52917629-d60bbf00-3313-11e9-9463-229434de25f4.png"  width="400"/>
<img src="https://user-images.githubusercontent.com/30471843/52917631-d99f4600-3313-11e9-94b6-b4e895625b2e.png"  width="400"/>

tfmorris commented:
Is there a way to reduce the copious amounts of whitespace in both the desktop and mobile layouts? Doesn't the current desktop put the title, etc, next to the cover image? Because of the aspect ratio of most covers (tall, narrow), there's a lot of space wasted if nothing is put next to them. 

For the mobile version, perhaps the "Read" button could be moved up there. We certainly don't need a full width Read button.

I'd also consider using different data elements on desktop and mobile. For example, is the publisher really important enough to take up valuable mobile screen real estate?

koderjoker commented:
@cdrini @tfmorris I'll create some UI prototypes with your respective suggestions to show how they look!

koderjoker commented:
Samples:

Reducing whitespace by shifting text to the right
![shift text](https://user-images.githubusercontent.com/30471843/53226538-52e6c200-36a1-11e9-9ac1-19bb244587c3.png)


Shifting download links to a dropdown (to overcome problem of whitespace and make download links bigger for mobile)
![download button](https://user-images.githubusercontent.com/30471843/53226628-a822d380-36a1-11e9-877e-47fc2c4c5e2f.png)

cdrini commented:
That looks _muuuch_ nicer! I think the download button is interesting, but in the interest of progress, let's drop it. I think it would take too long to make it look nice/work smoothly. Could you show a mockup with multiple books? Also place the DAISY link below the read button (so it's consistent with the desktop version of the site).

mekarpeles commented:
agree, the download button as designed is problematic for a variety of reasons including that the reading log button is green (conflicting color scheme)

I'd just add it as an option either under the read button, or as another option besides the other download formats.

koderjoker commented:
@mek Actually, I wasn't sure what component to use for a dropdown so I used the reading long button just to provide a rough sample of how a dropdown download button would look. 😆 Didn't intend for it to be used in the long run. 

I'm willing to design and implement a download button next if it is wanted. 

@cdrini @jdlrobson Mockup with multiple books on mobile
![screen shot 2019-02-28 at 2 48 10 pm](https://user-images.githubusercontent.com/30471843/53555441-f20b2e00-3b67-11e9-868e-748312f4cea1.png)

cdrini commented:
Awesome! That's coming along nicely. If we want a download button, I think we should investigate it in a new issue; for now let's not worry about it. Ok, some final changes:
There is still a little too much whitespace.
- Could you remove pretty much any vertical space? (Above the read button, below the download link, between edition entries).
- Could we have "worldcat" on the same line as "libraries near you?"
- would it be possible to have the download links on the same line as print disabled?
- increase the font size of the small text. Especially download/library links so that they are easier to hit on mobile.

Thanks for your patience, @koderjoker ! Assuming nothing goes horribly wrong, this should be the last set of changes.

koderjoker commented:
Firstly just summarising the structure of the table that is being reformatted, to make referring back to some stuff easier

For every tr
td1- Contains the book cover and title data
td2- Contains the read book button, daisy link and download links
td3- Contains library links
td4- Contains links for purchasing

With multiple books:
![screen shot 2019-03-01 at 10 27 18 pm](https://user-images.githubusercontent.com/30471843/53656321-f5daa580-3c78-11e9-9adf-eb285503d366.png)

Close up for mobile:
![screen shot 2019-03-01 at 10 31 04 pm](https://user-images.githubusercontent.com/30471843/53656325-f70bd280-3c78-11e9-9532-22678a01e8e1.png)

For mobile:
- Adjusted the padding as much as I could! Any other extra space is due to the absence of table data in the table row.
- "Libraries near you" could have multiple links associated with it (WorldCat, Library.link etc), so I think it would be best to have the links on the next line, to avoid the links wrapping to the next line
- Same reason as above, unless the daisy URL is shortened
- Increased the font size to make easier to hit the links!
- Removed the separator under the daisy link!

For tablet
![screen shot 2019-03-01 at 10 31 58 pm](https://user-images.githubusercontent.com/30471843/53656333-fb37f000-3c78-11e9-9ef5-d9e799efc80a.png)

Unfortunately, the tablet version still needs changes :p 

For starters, the width of the 'read eBook button' definitely needs to be reduced. We then could have the 'daisy and download links' on one line and the 'library and purchasing links' on another, but the read ebook button would still have space left over on the side... 

We could overcome that with a future download button though. The 'read ebook and download button' on one line, and 'library links and purchase links' on another.

jdlrobson commented on static/css/legacy-datatables.less:
```diff
@@ -30,7 +30,6 @@
   position: relative;
   z-index: @z-index-level-1;
   margin: 15px 0;
-  float: left;
```
Are you sure this doesn't break any other dataTables (I don't know the answer)?
If you're not confident, I'd suggest keeping this and adding a more specific CSS selector inside static/css/components/editions.less to override this rule.

koderjoker replied:
@jdlrobson this change is no longer necessary in the latest changes I made, so I'll leave it unchanged in the next commit 👍

jdlrobson commented:
This looks really good to me. I would be open to merging this as is provided we iterate on top of it. With that in mind I've removed the WIP tag ( please see https://github.com/internetarchive/openlibrary/issues/1339#issuecomment-468851785)

Great work @koderjoker it's great to start seeing these mobile experiences come to life!

koderjoker commented:
@jdlrobson @mekarpeles @cdrini @tfmorris 

Final look for 

Mobile
![screen shot 2019-03-03 at 12 32 12 am](https://user-images.githubusercontent.com/30471843/53686652-f7c36800-3d4f-11e9-8951-67f423cfc2c3.png)

Tablet
![screen shot 2019-03-03 at 1 02 38 am](https://user-images.githubusercontent.com/30471843/53686663-1164af80-3d50-11e9-9064-70fe47790b8d.png)

cdrini commented:
Nice @koderjoker! I know this has dragged on for a bit, but I think the design of this has improved greatly :) Here are some final notes/comments:

- Could you test this with a borrowable book that's been borrowed? Usually that would cause 2 buttons to appear: "Read" and "Return"; do they still display ok?
- Hmm, maybe try adding commas to separate the libraries (you can use CSS pseudo-elements to do this)? I think keeping them on one line is still better, even if there's a risk they'll wrap sometimes. An interface that is good 95% of the time and ok 5% of the time is better than an interface that is ok 100% of the time.
- Would it be possible to hide empty cells using the CSS `:empty` pseudo-selector?
- agreed the tablet view looks a little funky :P in the interest of progress, I'd suggest opening a new issue for that once this is merged.

mekarpeles commented:
I think download options and also Library link and worldcat should appear
above the called action button that way the call the action button is
always predictably the last item in the list and it feels like there is a
rhythm

On Sun, Mar 3, 2019, 9:38 AM Drini Cami <notifications@github.com> wrote:

> Nice @koderjoker <https://github.com/koderjoker>! I know this has dragged
> on for a bit, but I think the design of this has improved greatly :) Here
> are some final notes/comments:
>
>    - Could you test this with a borrowable book that's been borrowed?
>    Usually that would cause 2 buttons to appear: "Read" and "Return"; do they
>    still display ok?
>    - Hmm, maybe try adding commas to separate the libraries (you can use
>    CSS pseudo-elements to do this)? I think keeping them on one line is still
>    better, even if there's a risk they'll wrap sometimes. An interface that is
>    good 95% of the time and ok 5% of the time is better than an interface that
>    is ok 100% of the time.
>    - Would it be possible to hide empty cells using the CSS :empty
>    pseudo-selector?
>    - agreed the tablet view looks a little funky :P in the interest of
>    progress, I'd suggest opening a new issue for that once this is merged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/internetarchive/openlibrary/pull/1899#issuecomment-469045635>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA7tlZw1YqW580P3HQfoxc0BKJKmDTjhks5vTAixgaJpZM4a_t9->
> .
>

koderjoker commented:
>For every tr
td1- Contains the book cover and title data
td2- Contains the read book button, daisy link and download links
td3- Contains library links
td4- Contains links for purchasing

@mekarpeles do you mean that I should shift the 'Read eBook button' to the bottom of it's td and tr?

mekarpeles commented:
I think this makes sense (along with the print disabled Daisy link)

On Mon, Mar 4, 2019, 1:33 AM Kanchan Joshi <notifications@github.com> wrote:

> For every tr
> td1- Contains the book cover and title data
> td2- Contains the read book button, daisy link and download links
> td3- Contains library links
> td4- Contains links for purchasing
>
> @mekarpeles <https://github.com/mekarpeles> do you mean that I should
> shift the 'Read eBook button' to the bottom of it's td and tr?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/internetarchive/openlibrary/pull/1899#issuecomment-469182358>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA7tlfd52uZhzXF4oHmSVkqEapWa8Aogks5vTOhwgaJpZM4a_t9->
> .
>

cdrini commented:
(FYI, if the `tr` element is a flex box container, you can change the order of the cells using just css)

koderjoker commented:
Yup @cdrini, I'm aware of the above. Will implement the new changes as specified.

koderjoker commented:
@mekarpeles @cdrini @jdlrobson as per changes suggested


Single book sample
![screen shot 2019-03-06 at 12 27 22 am](https://user-images.githubusercontent.com/30471843/53830212-f12f2d80-3fa7-11e9-984a-94ff69d61d4f.png)


With multiple books
![screen shot 2019-03-06 at 12 37 41 am](https://user-images.githubusercontent.com/30471843/53830289-1c198180-3fa8-11e9-9fd8-d90caec7ad5b.png)

cdrini commented:
Sweet! Lgtm. Can we get this on dev, @mekarpeles ?

mekarpeles commented:
@cdrini + @koderjoker this is live on https://dev.openlibrary.org, let's test!

cdrini commented:
Oh, @koderjoker I don't think you've pushed your latest changes :P the thing on dev is what you had a few days ago.

koderjoker commented:
Wait, this wasn't supposed to happen :joy: I had put up the UI screenshots up first so that they could be approved before I pushed the latest changes! 
I'll push the changes once I get access to a computer!

jdlrobson commented:
> @mekarpeles @jdlrobson @cdrini encountering a couple of errors with the css build size
> 
> ```
> static/build/page-admin.css: 19.06KB > maxSize 19KB (gzip)
> static/build/page-form.css: 18.76KB > maxSize 18.7KB (gzip) 
> static/build/page-user.css: 18.62KB > maxSize 18.6KB (gzip)
> ```

These are all minor, so you should bump to them (but acknowledge the increases) in the commit message. You can find these values defined in package.json. While these values ideally would be going down, I think in this situation the increase is necessary, so you have my blessing to increase them :)

koderjoker commented:
UI issue for pages with multiple editions!

![screen shot 2019-03-07 at 11 39 42 pm](https://user-images.githubusercontent.com/30471843/53978978-88c08780-4132-11e9-9a34-e6fd6578822e.png)


Suggested changes:

![screen shot 2019-03-07 at 11 37 54 pm](https://user-images.githubusercontent.com/30471843/53979059-c02f3400-4132-11e9-8aef-7a1095fca6f0.png)
- Remove all extraneous floats used in the page (and then check to make sure it doesn't break other data tables)
- Shift number buttons below first, previous, next and last buttons for mobile devices
- Increase margin for certain elements

jdlrobson commented:
Love the suggested fix!!

mekarpeles commented:
(copied from my msg in slack: https://internetarchive.slack.com/archives/C0ETZV72L/p1551996974120600?thread_ts=1551902952.094300&cid=C0ETZV72L)

The `<tr>` with `border-bottom: #999` should probably be a lighter grey and also have a bit more padding-bottom.

Currently the “DAISY for print-disabled” is very cramped. It looks like `#editions tr .print-disabled-download` requires ~10px padding-bottom

re: border-bottom, perhaps something closer to hsla 0, 0, 87% (I think we have a light-grey in static/css/less/colors.less)

**before**

![screen shot 2019-03-07 at 14 19 34pmpst](https://user-images.githubusercontent.com/978325/53993402-28a2f480-40e4-11e9-99fe-d78724c10110.png)

**after**

![screen shot 2019-03-07 at 14 19 00pmpst](https://user-images.githubusercontent.com/978325/53993393-22147d00-40e4-11e9-9f8e-4a009d0735b6.png)

koderjoker commented:
Using flex order has affected the tab order. Got to remedy that too!

![tab order](https://user-images.githubusercontent.com/30471843/54015593-286b2d80-41a6-11e9-901d-11522f0ec380.gif)

mekarpeles commented:
This is really coming along, great work @koderjoker!

Is there a checklist of remaining todos? Or is this ready for review

koderjoker commented:
Ready for review!

tfmorris commented:
Did reducing with width of the "Read" button get lost? I commented on it a few weeks ago and @koderjoker said:

> For starters, the width of the 'read eBook button' definitely needs to be reduced. We then could have the 'daisy and download links' on one line and the 'library and purchasing links' on another, but the read ebook button would still have space left over on the side...

which sounded like agreement that it was too wide. Paradoxically, it's fine when the screen is wide, but below ~770 pix, it jumps to full width.

Also, a nit, but the `Library.link` styling looks weird/confusing. In the mobile rendering it looks like it's a modifier to WorldCat rather than a separate thing. Part this is because the label isn't capitalized correctly (ie. `Library.Link`).

Probably not for this ticket, but I wonder why we're sending any traffic to a commercial service which has a tiny fraction of the holdings of WorldCat. As long as we're keeping it though, we should at least make it clear to the users where we're sending them off to.

mekarpeles commented:
@tfmorris -- Library.Link is a team that spun out of oclc and they've been partnering directly with Internet Archive. Agreed re: presentation that opening a ticket would be great

re: full sized button, I also think this a good point -- may I also open this as a followup issue so this can finally be merged? 🎉

mekarpeles commented:
congrats @koderjoker !
